### DISCUSSION TITLE
MenuBuilder by tentacles

### DISCUSSION THREAD

**Discussion by Matsa59** [Omerlo-Technologies/squid#4] [Ideas]
Each tentacle should be able to define it's own menu using a MenuBuilder system.

## Syntax

We could create a behaviour that will create mandatory callback that could returns a list of items to add in the menu.

e.g.

```elixir
defmodule MyTentacle.MenuBuilder do
  def build do
    [
      live_redirect("homepage", to: "/"),
      link("homepage", to: "/")
      # ...
    ]
  end
end
```

However this syntax doesn't allow to add class based on a theme, so we will need to wrap each item.

We could also imagine something like

```elixir
defmodule MyTentacle.MenuBuilder do
  def build do
    [
      {:live_redirect, "homepage", to: "/"},
      {:link, "homepage", to: "/"}
      # ...
    ]
  end
end
```

Then in the `shell` (the "head" module that manage every tentacle) we could define a template that will act as a theme. e.g.

```elixir
defmodule ShellWeb.MenuBuilder do
  def render(assigns) do
    ~H"""
    <ul class="some-css-class">
      <%= for item in @items do %>
        <li class="some-css-li-class">
          <.render_menu_item(item) />
        </li>
      <% end %>
    </ul>
    """
  end
  
  def render_menu_item(menu_item) do
    # no need to explain here ...
  end
end
```

> Btw we maybe `head` is better name ?

## Sub menu

Each menu could "contain" other menu. e.g.

```
(menu A)
| item 1
| item 2
| (menu B)
| | - sub menu item 1
| | - sub menu item 2
```
> Here the menu from the tentacle A contains the menu of the tentacle B.

However, we don't want A to be dependent of B because it make more sense B depends to A.

**Comment by NatoBoram** [comment]
Any syntax for submenu items?

**Reply by Matsa59** [reply]
Sub menu will have the same syntax, the only problem that I have is : the menu should be registered by itself into his parent(s) menu. In that way child now their parent and not the reverse. I guess it will be a configuration, something like:

```elixir
config :tentacle_app,
  priority: 2,
  submenu_of: [:parent_tentacle_app_1, :parent_tentacle_app_2]
```

**Comment by Matsa59** [comment]
So after a lot of thinkings, we will create a new concept: `Partials`
The idea is creating a partial system that could be integrated to a more generic view

For example we could have a `MenuPartial` defined by multiples tentacles. Each partials will define a priority that will define the order for the lazy loading.

**Comment by Matsa59** [comment]
We could explain `Partial` as powerful `iframe`
